File: test/SIGNAL.erl
-module('SIGNAL').
-export([main/0,wait/3,loop/3,final/0]).
-define(BOOL(X), if (X) -> 1; true -> 0 end).


main() -> 
    wait(
    #{
        'GREEN_state' => 'off',
        'RED_state' => 'on'    },
    #{
        'SIGNAL_goahead' => 0,
        'SIGNAL_lg' => 0,
        'SIGNAL_lr' => 0
    },
    #{
        'SIGNAL_g' => 0,
        'SIGNAL_r' => 0,
        'GREEN_ok' => 0,
        'GREEN_error' => 0,
        'RED_error' => 0
    }).

final() ->
    erlister_rt:stop_timer(clk_GREEN_T1),
    erlister_rt:stop_timer(clk_RED_T2),
    ok.


wait(STATE,IN,OUT) ->
    receive
        {timeout,_Ref,_} -> loop(STATE,IN,OUT);
        INPUT -> loop(STATE,maps:merge(INPUT,IN),OUT)
    end.


loop(
    STATE = #{
                'GREEN_state' := ST_GREEN,
                'RED_state' := ST_RED
    },
    IN = #{
        'SIGNAL_goahead' := I_SIGNAL_goahead,
        'SIGNAL_lg' := I_SIGNAL_lg,
        'SIGNAL_lr' := I_SIGNAL_lr
    },
    OUT = #{
        'SIGNAL_g' := OUT_SIGNAL_g,
        'SIGNAL_r' := OUT_SIGNAL_r,
        'GREEN_ok' := OUT_GREEN_ok,
        'GREEN_error' := OUT_GREEN_error,
        'RED_error' := OUT_RED_error
    }) ->
    IN_SIGNAL_goahead = I_SIGNAL_goahead,
    IN_SIGNAL_lg = I_SIGNAL_lg,
    IN_SIGNAL_lr = I_SIGNAL_lr,
    IN_GREEN_request_on = I_SIGNAL_goahead,
    IN_GREEN_light_control = I_SIGNAL_lg,
    IN_GREEN_error_in = OUT_RED_error,
    CLK_GREEN_T1 = erlister_rt:read_timer(clk_SIGNAL_T1),
    ST1_GREEN = 
        case (ST_GREEN) of
        'off' ->
            if
                (not (IN_GREEN_error_in)) andalso (IN_GREEN_request_on) ->
                    erlister_rt:timer_start(clk_GREEN_T1,2000),
                    'on';
                true -> ST_GREEN
            end;
        'on' ->
            if
                (IN_GREEN_error_in) orelse (not (IN_GREEN_request_on)) ->
                    'off';
                true -> ST_GREEN
            end;
        _ -> ST_GREEN
        end,
    OUT1_GREEN_ok = ?BOOL(not (IN_GREEN_light_control) orelse (IN_GREEN_request_on)),
    OUT1_GREEN_error = ?BOOL((not (IN_GREEN_error_in)) andalso (((CLK_GREEN_T1 =:= timeout)) andalso ((not (IN_GREEN_light_control)) andalso (IN_GREEN_request_on)))),
    IN_RED_request_on = not (I_SIGNAL_goahead),
    IN_RED_light_control = I_SIGNAL_lr,
    IN_RED_error_in = OUT1_GREEN_error,
    IN_RED_gok = OUT1_GREEN_ok,
    CLK_RED_T2 = erlister_rt:read_timer(clk_SIGNAL_T2),
    ST1_RED = 
        case (ST_RED) of
        'off' ->
            if
                (IN_RED_error_in) orelse (IN_RED_request_on) ->
                    'on';
                true -> ST_RED
            end;
        'on' ->
            if
                (IN_RED_gok) andalso (not (IN_RED_request_on)) ->
                    erlister_rt:timer_start(clk_RED_T2,2000),
                    'off';
                true -> ST_RED
            end;
        _ -> ST_RED
        end,
    OUT1_RED_error = ?BOOL((not (IN_RED_error_in)) andalso (((CLK_RED_T2 =:= timeout)) andalso ((IN_RED_light_control) andalso ((ST1_RED =:= 'off'))))),
    OUT1_SIGNAL_g = ?BOOL((ST1_GREEN =:= 'on')),
    OUT1_SIGNAL_r = ?BOOL((ST1_RED =:= 'on')),
    wait(STATE#{
                'GREEN_state' => ST1_GREEN,
                'RED_state' => ST1_RED
    },
    IN,
    OUT#{
        'SIGNAL_g' => OUT1_SIGNAL_g,
        'SIGNAL_r' => OUT1_SIGNAL_r,
        'GREEN_ok' => OUT1_GREEN_ok,
        'GREEN_error' => OUT1_GREEN_error,
        'RED_error' => OUT1_RED_error
    }).

File: test/signal.c
typedef unsigned char  digital_t;
typedef unsigned short analog_t;
typedef unsigned long  timer_t;
typedef unsigned char  state_t;

enum {
    IN_SIGNAL_goahead,
    IN_SIGNAL_lg,
    IN_SIGNAL_lr,
    SIGNAL_NUM_INPUT
};

enum {
    OUT_SIGNAL_g,
    OUT_SIGNAL_r,
    OUT_GREEN_ok,
    OUT_GREEN_error,
    OUT_RED_error,
    SIGNAL_NUM_OUTPUT
};


enum {
    GREEN_off,
    GREEN_on
};

enum {
    RED_on,
    RED_off
};

typedef struct {
    digital_t input[SIGNAL_NUM_INPUT];
    digital_t output[SIGNAL_NUM_OUTPUT];
    timer_t   clk_GREEN_T1;
    timer_t   clk_RED_T2;
    state_t   GREEN_state;
    state_t   RED_state;
} SIGNAL_ctx_t;

extern int timer_init(timer_t* tp);
extern int timer_start(timer_t* tp);
extern int timer_stop(timer_t* tp);
extern int timer_timeout(timer_t* tp);

void init(SIGNAL_ctx_t* ctx)
{
    ctx->input[IN_SIGNAL_goahead] = 0;
    ctx->input[IN_SIGNAL_lg] = 0;
    ctx->input[IN_SIGNAL_lr] = 0;
    ctx->output[OUT_SIGNAL_g] = 0;
    ctx->output[OUT_SIGNAL_r] = 0;
    timer_init(&ctx->clk_GREEN_T1);
    timer_init(&ctx->clk_RED_T2);
    ctx->GREEN_state = GREEN_off;
    ctx->RED_state = RED_on;
}

void final(SIGNAL_ctx_t* ctx)
{
    timer_stop(&ctx->clk_GREEN_T1);
    timer_stop(&ctx->clk_RED_T2);
}

void machine(SIGNAL_ctx_t* ctx)
{
    digital_t SIGNAL_goahead;
    digital_t SIGNAL_lg;
    digital_t SIGNAL_lr;
    digital_t GREEN_request_on;
    digital_t GREEN_light_control;
    digital_t GREEN_error_in;
    digital_t RED_request_on;
    digital_t RED_light_control;
    digital_t RED_error_in;
    digital_t RED_gok;

    SIGNAL_goahead = ctx->input[IN_SIGNAL_goahead];
    SIGNAL_lg = ctx->input[IN_SIGNAL_lg];
    SIGNAL_lr = ctx->input[IN_SIGNAL_lr];
    GREEN_request_on = ctx->input[IN_SIGNAL_goahead];
    GREEN_light_control = ctx->input[IN_SIGNAL_lg];
    GREEN_error_in = ctx->output[OUT_RED_error];
    switch(ctx->GREEN_state) {
    case GREEN_off:
        if ((GREEN_error_in) || (!(GREEN_request_on))) {
            ctx->GREEN_state = GREEN_on;
            break;
        }
        break;
    case GREEN_on:
        if ((!(GREEN_error_in)) && (GREEN_request_on)) {
            ctx->GREEN_state = GREEN_off;
            timer_start(&ctx->clk_GREEN_T1);
            break;
        }
        break;
    default: break;
    }
    ctx->output[OUT_GREEN_ok] = !(GREEN_light_control) || (GREEN_request_on);
    ctx->output[OUT_GREEN_error] = (!(GREEN_error_in)) && ((timer_timeout(&ctx->clk_GREEN_T1)) && ((!(GREEN_light_control)) && (GREEN_request_on)));
    RED_request_on = !(ctx->input[IN_SIGNAL_goahead]);
    RED_light_control = ctx->input[IN_SIGNAL_lr];
    RED_error_in = ctx->output[OUT_GREEN_error];
    RED_gok = ctx->output[OUT_GREEN_ok];
    switch(ctx->RED_state) {
    case RED_off:
        if ((RED_gok) && (!(RED_request_on))) {
            ctx->RED_state = RED_on;
            timer_start(&ctx->clk_RED_T2);
            break;
        }
        break;
    case RED_on:
        if ((RED_error_in) || (RED_request_on)) {
            ctx->RED_state = RED_off;
            break;
        }
        break;
    default: break;
    }
    ctx->output[OUT_RED_error] = (!(RED_error_in)) && ((timer_timeout(&ctx->clk_RED_T2)) && ((RED_light_control) && ((ctx->RED_state == RED_off))));
    ctx->output[OUT_SIGNAL_g] = (ctx->GREEN_state == GREEN_on);
    ctx->output[OUT_SIGNAL_r] = (ctx->RED_state == RED_on);
}
